File: presences/src/main/resources/public/ts/services/ReasonService.ts
import {idiom as lang, ng} from 'entcore'
import { http, HttpResponse } from 'entcore-toolkit';
import {Reason, ReasonRequest} from "@presences/models/Reason";
import {REASON_TYPE_ID} from "@common/core/enum/reason-type-id";

export interface ReasonService {
    getReasons(structureId: string, reasonTypeId?: Number): Promise<Reason[]>;

    create(reasonBody: ReasonRequest, reasonTypeId?: Number): Promise<HttpResponse>;

    update(reasonBody: ReasonRequest): Promise<HttpResponse>;

    delete(reasonId: number): Promise<HttpResponse>;
}

export const reasonService: ReasonService = {
    getReasons: async (structureId: string, reasonTypeId?: Number): Promise<Reason[]> => {
        try {
            let reasonTypeIdParam: string = reasonTypeId != null && reasonTypeId != undefined ? `&reasonTypeId=${reasonTypeId}` : '';
            const {data} = await http.get(`/presences/reasons?structureId=${structureId}${reasonTypeIdParam}`);
            data.map((reason: Reason) => {
                if (reason.id === -1) {
                    reason.label = lang.translate(reason.label);
                    reason.hidden = true;
                }
            });
            return data;
        } catch (err) {
            throw err;
        }
    },

    create: async (reasonBody: ReasonRequest, reasonTypeId?: Number): Promise<HttpResponse> => {
        let reasonTypeIdParam: string = reasonTypeId ? `?reasonTypeId=${reasonTypeId}` : '';
        return http.post(`/presences/reason${reasonTypeIdParam}`, reasonBody);
    },

    update: async (reasonBody: ReasonRequest): Promise<HttpResponse> => {
        return http.put(`/presences/reason`, reasonBody);
    },

    delete: async (reasonId: number): Promise<HttpResponse> => {
        return http.delete(`/presences/reason?id=${reasonId}`);
    },
};

export const ReasonService = ng.service('ReasonService', (): ReasonService => reasonService);

File: app-presences/src/main/java/fr/openent/presences/service/ReasonService.java
package fr.openent.presences.service;

import fr.wseduc.webutils.Either;
import io.vertx.core.Future;
import io.vertx.core.Handler;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;

import java.util.List;

public interface ReasonService {

    /**
     * get reasons
     *
     * @param structureId  Structure Identifier
     * @param reasonTypeId Reason type Identifier
     */
    Future<JsonArray> get(String structureId, Integer reasonTypeId);

    /**
     * create reason
     *
     * @param reasonBody reasonBody fetched
     */
    Future<JsonObject> create(JsonObject reasonBody);

    /**
     * put reason
     *
     * @param reasonBody reasonBody fetched
     */
    Future<JsonObject> put(JsonObject reasonBody);

    /**
     * delete reason
     *
     * @param reasonId Reason Identifier
     * @param handler  Function handler returning data
     */
    void delete(Integer reasonId, Handler<Either<String, JsonObject>> handler);

    /**
     * get absence reasons from structure id and reason type id
     *
     * @param structureId Structure Identifier
     * @param handler     Function handler returning data
     */
    void fetchAbsenceReason(String structureId, Handler<Either<String, JsonArray>> handler);

    /**
     * get reasons from structure id and reason type id
     *
     * @param structureId  Structure Identifier
     * @param reasonTypeId Reason type Identifier
     * @return {@link Future} of {@link JsonObject}
     */
    Future<JsonArray> fetchReason(String structureId, Integer reasonTypeId);

    /**
     * get absence reasons from structure id and reason type id
     *
     * @param structureId Structure Identifier
     * @return {@link Future} of {@link JsonObject}
     */
    Future<JsonArray> fetchAbsenceReason(String structureId);

    /**
     * get absence reasons from structure i
     *
     * @param reasonIds Reason Identifier
     * @param handler   Function handler returning data
     */
    void getReasons(List<Integer> reasonIds, Handler<Either<String, JsonArray>> handler);
}
